cli.go

```diff
@@ -362,11 +362,25 @@ func printResult(out io.Writer, result *Result, mode DisplayMode, interactive, v
 		table.SetAlignment(tablewriter.ALIGN_LEFT)
 		table.SetAutoWrapText(false)
 
+		var forceTableRender bool
+		// This condition is true if statement is SelectStatement or DmlStatement
+		if verbose && len(result.ColumnTypes) > 0 {
+			forceTableRender = true
+			var headers []string
+			for _, field := range result.ColumnTypes {
+				typename := formatTypeSimple(field.GetType())
+				headers = append(headers, field.GetName()+"\n"+typename)
+			}
+			table.SetHeader(headers)
+		} else {
+			table.SetHeader(result.ColumnNames)
+		}
+
 		for _, row := range result.Rows {
 			table.Append(row.Columns)
 		}
-		table.SetHeader(result.ColumnNames)
-		if len(result.Rows) > 0 {
+
+		if forceTableRender || len(result.Rows) > 0 {
```

Comment:
Yes.
It is the originally proposed behavior in #190 , and I believe it is not harmful behavior.

There are interactive CLI SQL shells which print the header even if the result is empty. 

PostgreSQL

```
postgres=# SELECT 1 AS num LIMIT 0;
 num 
-----
(0 rows)
```

DuckDB

```
D SELECT 1 AS num LIMIT 0;
┌────────┐
│  num   │
│ int32  │
├────────┤
│ 0 rows │
└────────┘
```

Even mysql,  it is a wanted feature.

* https://stackoverflow.com/questions/32963136/how-can-i-retrieve-the-column-names-from-an-empty-mysql-select-query-result
* https://forums.mysql.com/read.php?10,225850

```
$ mysql --quick 

mysql> SELECT 1 AS num LIMIT 0;
+-----+
| num |
+-----+
+-----+
0 rows in set (0.00 sec)
``` 

IMO, the shape of the result shape is not obvious in many queries, so column name is useful even if the result is empty.
Repeated call with `DESCRIBE` is not easy because spanner-cli have not yet supported multiline histories.

Reply:
> It is the originally proposed behavior in #190 , and I believe it is not harmful behavior.

Yes, it's not a harmful behavior, but we need a consistency.

For users who're familiar with the current spanner-cli style (no headers for empty result), they will be also surprised with a different output style and they could even wonder if it's a bug or not.

So I think it might be better to use the current style (no headers for empty result) to keep the consistency across commands in spanner-cli.

Reply:
I feel it is the legacy of emulating `mysql` command.

Printing header have benefits over no headers for empty result, and I don't think it is needed to care users who relies the current behavior like parsing `Empty set` string.

> For users who're familiar with the current spanner-cli style (no headers for empty result), they will be also surprised with a different output style and they could even wonder if it's a bug or not.

I think it can be resolved by writing release notes.

If it is not acceptable as default verbose, we can add another more like [`-vv`](https://pkg.go.dev/github.com/jessevdk/go-flags#readme-example) or in ini file option.

Reply:
I think consistency should vary depending on perspective.
I feel that the current behavior of inconsistent header display is inconsistent.

> if user is eager to know the result shape regardless of the result set, they can use DESCRIBE SELECT command.

My opinion is that it is not easy to add `DESCRIBE` to previous ad-hoc query. https://github.com/cloudspannerecosystem/spanner-cli/pull/191#discussion_r1799620888
> Repeated call with DESCRIBE is not easy because spanner-cli have not yet supported multiline histories.

Reply:
> I feel that the current behavior of inconsistent header display is inconsistent.

If you feel that, we can think of two options regardless of `--verbose` usage.

1. Show header only for non-empty result
2. Show header anytime

I prefer option 1, but do you prefer option 2?

Reply:
Yes.
I think option 2 has more concrete benefits than option 1, and I believe that users will become more accepting of changes over time if the changes are documented in the release notes.

There is no standard for what should be displayed only when verbose. I personally intend to use verbose only and leave non-verbose as is.

Reply:
OK, then feel free to implement option 2 to keep consistency across commands in spanner-cli.

Reply:
> OK, then feel free to implement option 2 to keep consistency across commands in spanner-cli.

So, does that mean we should move towards always printing headers regardless of the `--verbose` flag or the type of statement?

* Not only `SelectStatement` and `DmlStatement`, all statement with results like `Show*Statement` will print header even if empty result
* Exception is completely no result statements (zero columns × zero rows ) like DDLs.